src/backend/pipeline/cont_adhoc.c

```diff
@@ -0,0 +1,1000 @@
+
+			CommitTransactionCommand();
+		}
+	}
+	PG_END_ENSURE_ERROR_CLEANUP(cleanup, PointerGetDatum(&cleanup_data));
+}
+
+/* 
+ * grabs the sql out of a statement like 
+ * select pipeline_exec_adhoc_query('select avg(x::int) from stream');
+ */
+static const char *
+get_stmt_sql(Node *node)
+{
+	/* assumes IsAdhocQuery has returned true on node already. */
+
+	SelectStmt *stmt = (SelectStmt *) (node);
+	Node *first_node = 0;
+	ResTarget *target = 0;
+	FuncCall *func = 0;
+	A_Const *sql_arg = 0;
+
+	first_node = (Node *) linitial(stmt->targetList);
+	target = (ResTarget *) (first_node);
+	func = (FuncCall *) (target->val);
+	sql_arg = linitial(func->args);
+
+	return sql_arg->val.val.str;
+}
+
+/* set a lower process priority for adhoc backends */
+static void
+set_nice_priority()
+{
+	int default_priority = getpriority(PRIO_PROCESS, MyProcPid);
+	int priority = Max(default_priority, 20 - ceil(continuous_query_proc_priority * (20 - default_priority)));
+
+	priority = nice(priority);
+}
+
+static void
+set_default_priority()
+{
+	/* TODO - figure out how to reset priority */
+}
+
+void
+ExecAdhocQuery(Node *stmt, const char *s)
+{
+	/* Break out of the top level xact (from exec_simple_query) */
+	AbortCurrentTransaction();
+
+	{
+		/* Set up our own top level mem ctx for cleanup */
+		MemoryContext adhoc_cxt = AllocSetContextCreate(CurrentMemoryContext,
+				"AdhocQueryContext",
+				ALLOCSET_DEFAULT_MINSIZE,
+				ALLOCSET_DEFAULT_INITSIZE,
+				ALLOCSET_DEFAULT_MAXSIZE);
+		
+		MemoryContext oldcontext = MemoryContextSwitchTo(adhoc_cxt);
+
+		/* This must be set so that newly created views are tagged properly */
+		SetAmContQueryAdhoc(true);
+
+		set_nice_priority();
+
+		/* At present, the queries always exit through the error path */
+		/* The 'errors' are from query cancel, or the frontend not responding to
+		 * a heartbeat */
+		PG_TRY();
+		{
+			List *parsetree_list;
+			const char *inner_sql = get_stmt_sql(stmt);
+			parsetree_list = pg_parse_query(inner_sql);
+
+			exec_adhoc_query(linitial(parsetree_list), inner_sql);
+		}
+		PG_CATCH();
+		{
+			SetAmContQueryAdhoc(false);
+
+			set_default_priority();
+			MemoryContextSwitchTo(oldcontext);
+			MemoryContextResetAndDeleteChildren(adhoc_cxt);
+			PG_RE_THROW();
+		}
+		PG_END_TRY();
+	}
+}
+
+/* returns true iff the function name matches the argument */
+static bool
+check_adhoc_func(FuncCall *func, const char *name)
+{
+	Node *node = 0;
+	Value *value = 0;
+	A_Const *sql_arg = 0;
+
+	if (list_length(func->funcname) != 1)
+		return false;
+
+	node = linitial(func->funcname);
+
+	if (!IsA(node, String))
+		return false;
+
+	value = (Value *) (node);
+
+	if (strcmp(value->val.str, name) != 0)
+		return false;
+
+	if (!func->args)
+		return false;
+
+	if (list_length(func->args) != 1)
+		return false;
+
+	sql_arg = linitial(func->args);
+
+	if (!IsA(sql_arg, A_Const))
+		return false;
+
+	return true;
+}
+
+/* returns true iff the query is an adhoc select */
+bool
+IsAdhocQuery(Node *node)
```

Comment:
@usmanm this is what we were originally going to do, but it turns out to be pretty complicated to do this cleanly within an actual function call. In particular, function calls don't have access to the client connection to write tuples to. We could probably find a way to work around that, but I don't think that it would be any cleaner than doing this signature-based intercept.